core/services/ocr2/plugins/ocr2keeper/evm21/encoding/interface.go

```diff
@@ -3,58 +3,56 @@ package encoding
 import (
 	"math/big"
 
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm21/mercury"
 	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg/v3/types"
 
 	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/automation_utils_2_1"
 	iregistry21 "github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/i_keeper_registry_master_wrapper_2_1"
 )
 
-type UpkeepFailureReason uint8
-type PipelineExecutionState uint8
```

Comment:
Dug more on this. The root cause of circular dependency is the Packer interface. The old Packer interface is very convoluted. For mercury, we just need a subset of the functions, so we defined a smaller Packer for mercury in mercury.go. 

```
type Packer interface {
	UnpackCheckCallbackResult(callbackResp []byte) (uint8, bool, []byte, uint8, *big.Int, error)
	PackGetUpkeepPrivilegeConfig(upkeepId *big.Int) ([]byte, error)
	UnpackGetUpkeepPrivilegeConfig(resp []byte) ([]byte, error)
	DecodeStreamsLookupRequest(data []byte) (*StreamsLookupError, error)
}
```
The first return of UnpackCheckCallbackResult is a PipelineExecutionState. If we keep using this encoding. PipelineExecutionState, then mercury package will have encoding and encoding package will also need mercury.

```
type Packer interface {
	UnpackCheckCallbackResult(callbackResp []byte) (uint8, bool, []byte, uint8, *big.Int, error)
...
```

Reply:
what's the plan to address this? We are causing a regression in typing

Reply:
In order to extract Mercury as a standalone lib in this PR, we changed the return type of `UnpackCheckCallbackResult` func from `encoding. PipelineExecutionState` to `uint8`, so as to avoid circular dependency btw `encoding` and `mercury`.

IMHO, this `encoding` [package](https://github.com/smartcontractkit/chainlink/tree/7f5c4325bbf2ea510485f27e07ffb1725d71220f/core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/encoding) can be refactored further.  Those [failure reasons and states](https://github.com/smartcontractkit/chainlink/blob/7f5c4325bbf2ea510485f27e07ffb1725d71220f/core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/encoding/interface.go#L17-L47) can be in a standalone package, while packer in its own package. Packer has one-way invocation to the `failure reason/states` package.